pkg/middleware/batch.go

```diff
@@ -0,0 +1,150 @@
+ ***************************************************************************/
+
+// Package middleware //
+package middleware
+
+import (
+	"bytes"
+	"encoding/json"
+	"github.com/go-chi/render"
+	"net/http"
+	"strings"
+	"time"
+)
+
+// BatchResposeItem holds the structure for each item
+type BatchResposeItem struct {
+	Status       int         `json:"status"`
+	RequestID    string      `json:"requestID"`
+	Method       string      `json:"method"`
+	URL          string      `json:"url"`
+	OrigResponse interface{} `json:"origResponse"`
+}
+
+// BatchResponse has the structure for the final response
+type BatchResponse struct {
+	StartedAt  time.Time          `json:"startedAt"`
+	EndedAt    time.Time          `json:"endedAt"`
+	ErrorCount int                `json:"errorCount"`
+	Response   []BatchResposeItem `json:"response"`
+}
+
+// BatchWriter implements http.ResponseWriter
+type BatchWriter struct {
+	statusCode int
+	headerMap  http.Header
+	method     string
+	url        string
+
+	BResponse BatchResponse
+}
+
+// WriteHeader sets the status code
+func (rec *BatchWriter) WriteHeader(code int) {
+	rec.statusCode = code
+}
+
+// Write is just the collector for BatchResponse
+func (rec *BatchWriter) Write(b []byte) (int, error) {
+
+	var data interface{}
+	if err := json.Unmarshal(b, &data); err != nil {
+		return 0, err
+	}
+	if rec.statusCode != http.StatusOK {
+		rec.BResponse.ErrorCount++
+	}
+	requestID := ""
+	if header, ok := rec.headerMap["X-Request-Id"]; ok {
+		requestID = header[0]
+
+	}
+	bri := BatchResposeItem{Status: rec.statusCode, OrigResponse: data, Method: rec.method, URL: rec.url, RequestID: requestID}
+	rec.BResponse.Response = append(rec.BResponse.Response, bri)
+	rec.BResponse.EndedAt = time.Now()
+	return 0, nil
+}
+
+// Header returns header map
+func (rec *BatchWriter) Header() http.Header {
+	return rec.headerMap
+}
+
+// Request is the original request that is used for batching
+type Request struct {
+	Operations []struct {
+		Method      string                 `json:"method"`
+		URL         string                 `json:"url"`
+		OperationID string                 `json:"operationID"`
+		Body        map[string]interface{} `json:"body"`
+		Params      map[string]string      `json:"params"`
+		Headers     map[string]string      `json:"headers"`
+	} `json:"operations"`
+}
+
+// BatchRouter intercepts requests for the given url to return a StatusOK.
+func BatchRouter(next http.Handler) http.Handler {
+	fn := func(w http.ResponseWriter, r *http.Request) {
+		if r.Method == "POST" && strings.HasSuffix(strings.ToLower(r.URL.Path), "/batch") {
+
+			decoder := json.NewDecoder(r.Body)
+			var req Request
+			err := decoder.Decode(&req)
+			if err != nil {
+				render.JSON(w, r, `{"error": "cannot decode the operation body"}`)
+				return
+			}
+			batchWriter := BatchWriter{200, make(http.Header), "", "",
+				BatchResponse{StartedAt: time.Now(), Response: []BatchResposeItem{}}}
```

Comment:
Having a single global `BatchWriter` like this is going to be problematic and prevent us from extending to add concurrency support. Since we're reusing the statusCode, method, url and headers they can easily collide from one request to another. Alternatively we should construct a _new_ writer (collector) for each operation and _append_ the contents onto the BatchResponse instance.

The pseudo code would be something like:
```go
br := BatchResponse{StartedAt: time.Now(), Response: []BatchResposeItem{}}}
for _, op := range req.Operations {
	// Construct collector
	col := ResponseCollector{StartedAt: time.Now()}

	// Build request
	bytesBody, e := json.Marshal(op.Body)
	...

	// Make request
	next.ServeHTTP(col, opReq)
	col.EndedAt = time.Now()

	// Append response item
	br.append(col)
}
br.EndedAt = time.Now()
render.JSON(w, r, br)
```